libdrgn/python/helpers.c

```diff
@@ -249,3 +250,267 @@ PyObject *drgnpy_linux_helper_pgtable_l5_enabled(PyObject *self, PyObject *args,
 		return PyErr_Format(PyExc_ValueError, "not Linux kernel");
 	Py_RETURN_BOOL(prog->prog.vmcoreinfo.pgtable_l5_enabled);
 }
+
+static void GenericIterator_dealloc(GenericIterator *self)
+{
+	if (self->iterator) {
+		self->iterator_deinit(self->iterator);
+		free(self->iterator);
+	}
+	Py_XDECREF(self->prog);
+	Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static PyObject *GenericIterator_next(GenericIterator *self)
+{
+	return self->next(self);
+}
+
+PyTypeObject GenericIterator_type = {
+	PyVarObject_HEAD_INIT(NULL, 0)
+	.tp_name = "_drgn._GenericIterator",
+	.tp_basicsize = sizeof(GenericIterator),
+	.tp_dealloc = (destructor)GenericIterator_dealloc,
+	.tp_flags = Py_TPFLAGS_DEFAULT,
+	.tp_iter = PyObject_SelfIter,
+	.tp_iternext = (iternextfunc)GenericIterator_next,
+};
+
+static PyObject *for_each_task_next(GenericIterator *self)
+{
+	struct drgn_error *err;
+	struct task_iter *iterator = (struct task_iter *)self->iterator;
+	struct drgn_object *entry;
+	err = task_iter_next(iterator, &entry);
+	if (err)
+		return set_drgn_error(err);
+	if (!entry)
+		return NULL;
+	DrgnObject *ret = DrgnObject_alloc(self->prog);
+	if (!ret)
+		return NULL;
+	err = drgn_object_copy(&ret->obj, entry);
+	if (err)
+		return set_drgn_error(err);
+	return (PyObject *)ret;
+}
+
+GenericIterator *drgnpy_linux_helper_for_each_task(PyObject *self,
+						   PyObject *args,
+						   PyObject *kwds)
+{
+	static char *keywords[] = {"prog_or_ns", NULL};
+	struct drgn_error *err = NULL;
+	struct prog_or_ns_arg prog_or_ns;
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&:for_each_task",
+					 keywords, &prog_or_pid_ns_converter,
+					 &prog_or_ns))
+		return NULL;
+
+	GenericIterator *iterator =
+		(GenericIterator *)GenericIterator_type.tp_alloc(
+			&GenericIterator_type, 0);
+	if (!iterator)
+		goto out;
+	iterator->prog = NULL;
+	iterator->next = for_each_task_next;
+	iterator->iterator_deinit = (void (*)(void *))task_iter_deinit;
+	iterator->iterator = malloc(sizeof(struct task_iter));
+	if (!iterator->iterator) {
+		err = &drgn_enomem;
+		goto out;
+	}
+	err = task_iter_init(iterator->iterator, prog_or_ns.ns);
+out:
+	prog_or_ns_cleanup(&prog_or_ns);
+	if (err || !iterator) {
+		Py_XDECREF(iterator);
+		return err ? set_drgn_error(err) : NULL;
+	}
+	iterator->prog = prog_or_ns.prog;
+	Py_INCREF(iterator->prog);
+	return iterator;
```

Comment:
Allso not a fan of the `out` label handling every case here. I'd prefer something like:

```suggestion
	static char *keywords[] = {"prog_or_ns", NULL};
	struct drgn_error *err;
	struct prog_or_ns_arg prog_or_ns;

	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&:for_each_task",
					 keywords, &prog_or_pid_ns_converter,
					 &prog_or_ns))
		return NULL;

	GenericIterator *iterator =
		(GenericIterator *)GenericIterator_type.tp_alloc(
			&GenericIterator_type, 0);
	if (!iterator)
		goto out;
	iterator->prog = prog_or_ns.prog;
	Py_INCREF(iterator->prog);
	iterator->next = for_each_task_next;
	iterator->iterator_deinit = (void (*)(void *))task_iter_deinit;
	iterator->iterator = malloc(sizeof(struct task_iter));
	if (!iterator->iterator) {
		PyErr_NoMemory();
		Py_DECREF(iterator);
		iterator = NULL;
		goto out;
	}
	err = task_iter_init(iterator->iterator, prog_or_ns.ns);
	if (err) {
		set_drgn_error(err);
		Py_DECREF(iterator);
		iterator = NULL;
		goto out;
	}
out:
	prog_or_ns_cleanup(&prog_or_ns);
	return iterator;
```

WIth the same idea for the other iterators.